File: cpp/GaleShapley.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <string>
#include <map>

using namespace std;

vector<pair<string, string>> galeShapley(
    map<string, vector<string>>& male_prefs,
    map<string, vector<string>>& female_prefs
) {
    map<string, vector<string>> mp = male_prefs;
    map<string, pair<map<string, int>, pair<int, string>>> fp;
    map<string, int> curr_pair;

    for (auto& p : female_prefs) {
        fp[p.first] = {map<string, int>(), {-1, ""}};
        fp[p.first].first = map<string, int>();
        for (int i = 0; i < p.second.size(); i++) {
            fp[p.first].first[p.second[i]] = i;
        }
    }
    
    for (auto& p : male_prefs) {
        curr_pair[p.first] = 0;
    }

    while (mp.size() > 0) {
        map<string, vector<string>> next;
        for (auto& p : mp) {
            string male = p.first;
            int pointer = curr_pair[male];
            string female = p.second[pointer];

            if (fp[female].second.first == -1) {
                fp[female].second = {fp[female].first[male], male};
            } else if (fp[female].second.first != -1 && fp[female].first[male] < fp[female].second.first) {
                string free_male = fp[female].second.second;
                next[free_male] = male_prefs[free_male];
                curr_pair[free_male]++;
                fp[female].second = {fp[female].first[male], male};
            } else {
                next[male] = male_prefs[male];
                curr_pair[male]++;
            }
        }
        mp = next;
    }
    vector<pair<string, string>> pairings;
    for (auto& p : male_prefs) {
        pairings.push_back({p.first, p.second[curr_pair[p.first]]});
    }
    return pairings;
}

int main() {
    map<string, vector<string>> male_prefs;
    map<string, vector<string>> female_prefs;

    male_prefs["A"] = {"O", "M", "N", "L", "P"};
    male_prefs["B"] = {"P", "N", "M", "L", "O"};
    male_prefs["C"] = {"M", "P", "L", "O", "N"};
    male_prefs["D"] = {"P", "M", "O", "N", "L"};
    male_prefs["E"] = {"O", "L", "M", "N", "P"};

    female_prefs["L"] = {"D", "B", "E", "C", "A"};
    female_prefs["M"] = {"B", "A", "D", "C", "E"};
    female_prefs["N"] = {"A", "C", "E", "D", "B"};
    female_prefs["O"] = {"D", "A", "C", "B", "E"};
    female_prefs["P"] = {"B", "E", "A", "C", "D"};

    vector<pair<string, string>> pairings = galeShapley(male_prefs, female_prefs);
    cout << "Derived pairings:" << endl;
    for (auto& p : pairings) {
        cout << p.first << "<->" << p.second << endl;
    }
    return 0;
}

File: python/misc/GaleShapley.py
def gale_shapley(male_prefs: list, female_prefs: list) -> list[tuple]:
    mp = {male: pref for male, pref in male_prefs}

    males = male_prefs
    # We hash the index pointer for each male to kee track of their curr/ex engagement, increments by one if they are rejected for another
    # male
    curr_pair = {male: 0 for male, _ in male_prefs}
    fp = {}

    # -1 indicates the female is free, we hash the rankings of their preferences for easier reference and save the current male
    # they are engaged to
    for female, pref in female_prefs:
        fp[female] = [{male: rank for rank, male in enumerate(pref)}, (-1, None)]

    iteration = 0
    while males:
        print(f"Iteration {iteration}")

        temp = []
        for male, pref in males:
            pointer = curr_pair[male]
            female = pref[pointer]
            # If the female is free, we simply assign her to that male
            if fp[female][1][0] == -1:
                fp[female][1] = (fp[female][0][male], male)
            # If she isn't free and the current male is preferred, reject her current engagement and replace him
            elif fp[female][1][0] != -1 and fp[female][0][male] < fp[female][1][0]:
                free_male = fp[female][1][1]
                temp.append((free_male, mp[free_male]))
                curr_pair[free_male] += 1
                fp[female][1] = (fp[female][0][male], male)
            # Other add the rejected male to the list and go down his preference list
            else:
                temp.append((male, mp[male]))
                curr_pair[male] += 1

        males = temp
        iteration += 1
        print(f"Current pairings: {[(fp[fem][1][1], fem) for fem in fp if fp[fem][1][0] != -1]}")
    
    pairings = [(male, pref[curr_pair[male]]) for male, pref in male_prefs]
    print("Derived pairings:")
    for pair in pairings:
        print(pair)
    return pairings

if __name__ == '__main__':
    male_prefs = [
        ('A', tuple("OMNLP")),
        ('B', tuple("PNMLO")),
        ('C', tuple("MPLON")),
        ('D', tuple("PMONL")),
        ('E', tuple("OLMNP")),
    ]
    female_prefs = [
        ('L', tuple("DBECA")),
        ('M', tuple("BADCE")),
        ('N', tuple("ACEDB")),
        ('O', tuple("DACBE")),
        ('P', tuple("BEACD")),
    ]
    pairings = gale_shapley(male_prefs, female_prefs)
